contrib/endpoints/src/api_manager/check_security_rules.cc

```diff
@@ -14,92 +14,339 @@
+              } else {
+                pchecker->env_->LogError(std::string("Test API failed with ")
+                    + status.ToString());
+                status = Status(Code::INTERNAL, kFailedFirebaseTest);
+              }
+
+              continuation(status);
+            });
+}
+
+Status AuthzChecker::ParseReleaseResponse(const std::string &json_str,
+                                          std::string *ruleset_id) {
+  grpc_json *json = grpc_json_parse_string_with_len(
+      const_cast<char *>(json_str.data()), json_str.length());
+
+  if (!json) {
+    return Status(Code::INVALID_ARGUMENT, kInvalidResponse);
+  }
+
+  Status status = Status::OK;
+  const char *id = GetStringValue(json, "rulesetName");
+  (*ruleset_id) = (id == nullptr) ? "" : GetStringValue(json, "rulesetName");
+
+  if (ruleset_id->empty()) {
+    env_->LogError("Empty ruleset Id received from firebase service");
+    status = Status(Code::INTERNAL, kInvalidResponse);
+  } else {
+    env_->LogInfo(std::string("Received ruleset Id: ") + *ruleset_id);
+  }
+
+  grpc_json_destroy(json);
+  return status;
 }
 
-const std::string &AuthzChecker::GetAuthToken() {
-  // TODO: Get Auth token for accessing Firebase Rules service.
-  static std::string empty;
-  return empty;
+Status AuthzChecker::ParseTestResponse(
+    std::shared_ptr<context::RequestContext> context,
+    const std::string &json_str) {
+  grpc_json *json = grpc_json_parse_string_with_len(
+      const_cast<char *>(json_str.data()), json_str.length());
+
+
+  if (!json) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "Invalid JSON response from Firebase Service");
+  }
+
+  Status status = Status::OK;
+  Status invalid = Status(Code::INTERNAL, kInvalidResponse);
+
+  const grpc_json *testResults = GetProperty(json, "testResults");
+  if (testResults == nullptr) {
+    env_->LogError("TestResults are null");
+    status = invalid;
+  } else {
+    const char *result = GetStringValue(testResults->child, "state");
+    if (result == nullptr) {
+      env_->LogInfo("Result state is empty");
+      status = invalid;
+    } else if (std::string(result) != "SUCCESS") {
+      status = Status(Code::PERMISSION_DENIED,
+              std::string("Unauthorized ")
+              + context->request()->GetRequestHTTPMethod()
+              + " access to resource " + context->request()->GetRequestPath(),
+              Status::AUTH);
+    }
+  }
+
+  grpc_json_destroy(json);
+  return status;
+}
+
+std::string AuthzChecker::GetOperation(std::string httpMethod) {
+  if (httpMethod == "POST") {
+    return std::string("create");
+  }
+
+  if (httpMethod == "GET" || httpMethod == "HEAD" || httpMethod == "OPTIONS") {
+    return std::string("get");
+  }
+
+  if (httpMethod == "DELETE") {
+    return std::string("delete");
+  }
+
+  return std::string("update");
+}
+
+std::string AuthzChecker::BuildTestRequestBody(
+    std::shared_ptr<context::RequestContext> context) {
+
+  std::ostringstream ss;
```

Comment:
There are multiple proto files and this will come from the firebase codebase. I am not sure if we can just pick up the code and put it in our codebase. I will need to talk to Limin / Wencheng and may be even Firebase guys before doing this.